catalogue_of_life/Loader.java

```diff
@@ -0,0 +1,242 @@
+        TsvParserSettings settings = new TsvParserSettings();
+        settings.setHeaderExtractionEnabled(true);
+        settings.setMaxCharsPerColumn(1_000_000);
+
+        TsvParser parser = new TsvParser(settings);
+        InputStream inputStream = Files.newInputStream(tsvFile);
+        parser.beginParsing(inputStream);
+
+        return parser;
+    }
+
+    private static TsvWriter createTsvWriter(Path tsvFile) throws IOException {
+        TsvWriterSettings writerSettings = new TsvWriterSettings();
+        writerSettings.setHeaderWritingEnabled(true);
+        writerSettings.setMaxCharsPerColumn(1_000_000);
+
+        return new TsvWriter(Files.newOutputStream(tsvFile), writerSettings);
+    }
+
+    private static boolean isValidMarineRegionID(String areaID) {
+        return isNumeric(areaID);
+    }
+
+    private static boolean isNumeric(String s) {
+        try {
+            Long.parseLong(s);
+        } catch (NumberFormatException e) {
+            return false;
+        }
+        return true;
+    }
+
+    public static void loadData(Path dataDirectory, Options options) throws IOException {
+        extractSchema(dataDirectory);
+
+
+        LoadOptions loadOptions = new LoadOptions();
+        loadOptions.dataConfigFilePath = createTypeDBLoaderConfig(dataDirectory).toString();
+        loadOptions.databaseName = "catalogue-of-life";
+        loadOptions.typedbURI = options.typedbURI();
+        loadOptions.cleanMigration = true;
+        loadOptions.loadSchema = false;
+
+        TypeDBLoader loader = new TypeDBLoader(loadOptions);
+        loader.load();
+    }
+
+    public static void extractSchema(Path dataDirectory) throws IOException {
+        InputStream schema = Loader.class.getResourceAsStream("/catalogue_of_life/schema.tql");
+        if (schema == null) throw new FileNotFoundException("schema.tql not found within the jar!");
+        Files.copy(schema, dataDirectory.resolve("schema.tql"), StandardCopyOption.REPLACE_EXISTING);
+    }
+
+    public static Path createTypeDBLoaderConfig(Path dataDirectory) throws IOException {
+        Path config = dataDirectory.resolve("loader-config.json");
+
+        try (FileWriter configWriter = new FileWriter(config.toFile())) {
+            InputStream template = Loader.class.getResourceAsStream("/catalogue_of_life/loader-config.json.template");
+            if (template == null) throw new FileNotFoundException("loader-config.json.template not found within the jar!");
+
+            BufferedReader reader = new BufferedReader(new InputStreamReader(template));
+            String replacement = dataDirectory.toString();
+            while (reader.ready()) {
+                String line = reader.readLine().replace("${data-directory}", replacement);
+                configWriter.write(line + "\n");
+            }
+        }
+
+        return config;
+    }
+
+    @CommandLine.Command(name = "loader", mixinStandardHelpOptions = true)
+    public static class Options {
```

Comment:
what do you think, maybe we can move this to a sibling class called `CLIOptions`?

Reply:
It's tiny, and I quite like the scoping of `...Loader.Options` which instantly tells you exactly what it's for. `Loader.CLIOptions` is probably better though.